---
title: "A3: Data set Pathway and Network Analysis"
author: "Yining Ding"
output:
  html_notebook:
    toc: yes
    toc_depth: 3
bibliography: A2.bib
---
# Data source 
* The GEO series used in this assignment is GSE84054, "Transcriptome profiling of ER+ breast cancer primary tumor and its tumorsphere derivative" [@goh2017chromosome]. The data contains 12 patients, each having both tumursphere and primary tumour. From the paper, the authors performed RNA-seq analysis to find differentially expressed genes from the two groups. They also discovered a potential drug target. The drug they sugguested that may have effect is Pacritinib which is still being annotated.


# Objective of A3
The objective of this third assignment is to use the ranked file from A2 and perform non-thresholded analysis with GSEA (Version 4.0.3) [@GSA]. Then the result is compared to the result from thresholded analysis in A2. Finally, cytoscape (Version 3.7.2, Java 1.8.0_162 by Oracle Corporation)[@lopes2010cytoscape] is used to help to create a visualization of the up and down regulated genes along with annotations from the Bader lab. Post analysis is performed with drugBank datasets from Bader lab as well. Other details such as performing the analysis of cytoscape can be retrieved from [my Journal](https://github.com/bcb420-2020/student_helen307/wiki/a3)

# A1 and A2 
Work from A1 and A2 are embedded in this notebook. In A1, we performed data cleaning, normalization and general analysis with density and box plots. In A2, we explored the differentially expressed genes and ranked them with thresholded analysis to highlight the top terms. And also found some evidence that supported my work from some other papers. Other details from A1 and A2 can also be found in my journal.


```{r child='A2.Rmd', include=FALSE, eval=TRUE, echo=FALSE}
```

# A3

R Libraries used: 

* RCurl[@RCurl]
* knitr[@knitr]
* GSA[@GSA]

```{r, include=FALSE}

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("RCurl")

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("GSA")

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("knitr")

library(RCurl)
library(GSA)
library(knitr)
```
# Part 1: Non-thresholded Gene set Enrichment Analysis

## STEP 1: Get Bader lab gene sets
* The code is inspired by lecture notes: non-thresholded analysis unit.
```{r}
gmt_url = "http://download.baderlab.org/EM_Genesets/current_release/Human/symbol/"
# list all the files on the server
filenames = getURL(gmt_url)
tc = textConnection(filenames)
contents = readLines(tc)
close(tc)
# get the gmt that has all the pathways and does not include terms inferred from
# electronic annotations(IEA) start with gmt file that has pathways only
rx = gregexpr("(?<=<a href=\")(.*.GOBP_AllPathways_no_GO_iea.*.)(.gmt)(?=\">)", contents, 
    perl = TRUE)
gmt_file = unlist(regmatches(contents, rx))
download.file(paste(gmt_url, gmt_file, sep = ""), destfile = dest_gmt_file)
```

## STEP 2: Compute ranks 
* I obtained the rank file, computed the ranks and sorted table by ranks. The output file is "expr_RNAseq_ranks.rnk" which is to be loaded into GSEA for further analysis.

```{r}
# compute ranks
qlf_output_hits_withgn[,"rank"] <- log(qlf_output_hits_withgn$PValue, base = 10) * sign(qlf_output_hits_withgn$logFC)

# sort table by ranks
qlf_output_hits_withgn <- qlf_output_hits_withgn[order(qlf_output_hits_withgn$rank),]

# write gene name and rank to table
write.table(x=qlf_output_hits_withgn[, c(2, ncol(qlf_output_hits_withgn))],
            file="expr_RNAseq_ranks.rnk",sep = "\t",
            row.names = FALSE,col.names = FALSE,quote = FALSE)
```



## STEP 3: Conduct non-thresholded gene set enrichment analysis

I used the computed ranked set of genes to compute the non-thresholded gene set entichment analysis with GSEA_4.0.3.[@GSA] I loaded the rank file "expr_RNAseq_ranks.rnk" and Bader Lab gene set "human_GOBP_AllPathways_no_GO_iea_March_01_2020_symbol.gmt". The parameters are permutation = 100, no collapse, max = 200, min = 15. The file obtained is: "A3_part1_non_thres.GseaPreranked.1584469398976". Here is a screenshot before performing the analysis on GSEA.
```{r, echo=FALSE}
knitr::include_graphics("/Users/helending/Documents/bcb420/image/gsea_before.png")
```


## STEP 4: Summarize enrichment results

My data is divided into "POS" and "NEG" which correspond to up-regulatedand down-regulated genes. 

* For the "POS" group, humoral immune response is the top term with a p-val of 0, 39 genes in its leading edge, and the top gene associated with this geneset is IGKV2D-28(probe name). 
* In the "NEG" group, TGF beta receptor is the top term, with a p-val of 0, 185 genes in its leading edge and the top gene associated with is PRKCB.


## STEP 5: Compare to the results from thresholded analysis

I used g:profiler in A2 to conduct thresholded analysis. The result showed that both the up-regulated and the down-regulated genes have metabolic terms associated. Whereas using the non-thresholded methods, immune response seems to be associated with up-regulated genes. Beta-receptor seems to be associated with the down-regulated group which is the same as the result from g:profiler. Both thresholded and non-thresholded over-representation analysis shows that the disease is associated with immune response and metabolic malfunctions. Such result is consistent with the result from the paper, and I have found some of the evidences from the other papers in A2, showing that breast cancer is associated with these top terms.


# Part 2: Visualize Gene set Enrichment Analysis

## STEP 1: Create an enrichment map

* The following visualization is conducted using Cytoscape [@lopes2010cytoscape]
* The parameters I used when creating the enrichment map: FDR was set to 0.01, others are kept as default.
* Basic information: there are 456 nodes and 5087 edges in the resulting map.
* Here is a screenshot that gives information about the creation of the enrichment map.

```{r, echo=FALSE}
knitr::include_graphics("/Users/helending/Documents/bcb420/image/creation_enrichment_map.png")
```


## STEP 2: Annotate the enrichment map

I defined the main biological themes by using clicking on "auto annotate" -> "annotate". The system created circles around each cluster which corresponds to the most frequent node lavles in the cluster.
* Here is a screenshot showing that each cluster is given a name.


```{r, out.width = "600px", echo=FALSE}
knitr::include_graphics("/Users/helending/Documents/bcb420/image/cytoscape_circle_example.png")
```

## STEP 3: Manually edit the network

* I made sure that there is no overlap between the labels and the nodes. I also oragnized them to make the interactions clearer. The legend is obtained from Bader lab's website. 
* I did not screenshot the network since it is the same as the resulting graph that also contains the two drugs I picked. The screenshot can be found in the section: " Post analysis - Drug Bank".
* **Note: In order to keep the image in a high resolution, I chose to attach the pdf (a png file does not have high resolution). Please download to have a better view.**

## STEP 4: Collapse to a theme network
* Generally, the major themes I found after collapsing the annotated network: immune responses, cell proliferation, tageting, and signaling.There seems to be no novel pathways. 
* Here is an overview of the collapsed graph.

```{r, echo=FALSE}
include_graphics("/Users/helending/Documents/bcb420/image/cropped_no_drug_collapsed.pdf")
```
## STEP 5: Interpretation of the annotated network
* Terms on the left are up-regulated genes and terms on the right are down-regulated genes. The lines shows us the interactions between each most frequent terms.
* On the left side, we can see that the term that has the most interactions with the rest is "response immune immunoglobulin". It is connected to terms like "lymphocyte proliferation", "differentiation", "negative cell regulation", "signaling", etc. Elevated level immunoglobulin antibodies is frequently observed in cancers of epithelial origin, including carcinomas of breast, colon, and liver.[@qiu2003human] Therefore, it justifies the reason why this term is up-regulated.
* On the right side, we notice the main term is "APC dedegradation" (with the most genes). Antigen presenting cells (APCs) are immune cells that specialize in presenting an antigen to a T-cell. APC protein acts as a tumor suppressor, therefore, the degradation of APC means that the suppressor is no longer fully supplied which can not stop the disease from expanding. Other terms are related to cell * transport, etc.
* In conclusion, those terms are predicted by the g:profiler in A2 even if they don't have the exact same names. Therefore, we can conclude that the disease is mostly due to immune response malfunction, which then leads to problems in cell metabolism in various ways.

# STEP 6: Post analysis - Drug Bank

The reason why I choose to do the post analysis with drugs is because the paper mentioned about finding a target for a drug. The drug is Pacritinib. However, this drug is not annotated, and thus not stored in the Bader Lab file. Therefore, I chose the top drug (Abciximab) [@law2014drugbank] from Bader lab approved drugs file. The paper also mentioned that they use Gemcitabine to treat breast cancer if the disease does not expand further. 

* Parameters: Mann Whiteney(One-sided greater): since I do not care whether it targets the up-regulated or down-regulated genes.

* The resulting graph shows that there are no connection for Gemcitabine with any pathways, and a lot of up-regulated genes are targeted by Abciximab. The terms that are targeted are mostly the up-regulated genes, such as "response immune immunoglobin", to reduce the elevated level of the antibody. It also reduces the level of other cell metabolic terms that are in the same pathway as the "response immune immunoglobin" (discussed in the section above). The main function of this drug is to reduce antibodies, and other over-produced molecules, but not elevating the weak terms (the down-regulated genes). 

* The only thing not expected is that Gemcitabine does not target the term "cell proliferation positive" which is what the main object of Gemcitabine. Gemcitabine is a drug that kills the fast-growing cells. Therefore, it is not expected to see that this drug does nothing in this disease pathway.

* It might be hard to see the Gemcitabine, it is right beside Abciximab,.
```{r, echo=FALSE}
knitr::include_graphics("/Users/helending/Documents/bcb420/image/publication_figure.png")
```

```{r fig.height=2.7, fig.width=9, echo=FALSE}
include_graphics("/Users/helending/Documents/bcb420/image/cropped_no_drug_collapsed.pdf")
```
# References
